lib/Doctrine/DBAL/Schema/Comparator.php

```diff
@@ -432,12 +433,12 @@ public function diffColumn(Column $column1, Column $column2)
             $changedProperties[] = 'comment';
         }
 
-        if ($properties1['default'] !== $properties2['default'] ||
-            // Null values need to be checked additionally as they tell whether to create or drop a default value.
-            // null != 0, null != false, null != '' etc. This affects platform's table alteration SQL generation.
-            ($properties1['default'] === null && $properties2['default'] !== null) ||
-            ($properties2['default'] === null && $properties1['default'] !== null)
-        ) {
+        $default1IsNull = $properties1['default'] === null;
+        $default2IsNull = $properties2['default'] === null;
+
+        // Null values need to be checked additionally as they tell whether to create or drop a default value.
+        // null != 0, null != false, null != '' etc. This affects platform's table alteration SQL generation.
+        if ($default1IsNull !== $default2IsNull || strcmp($properties1['default'], $properties2['default']) !== 0) {
```

Comment:
From my understanding, the original problem was that `['default']` could be i.e. integer (when defined in ORM for example), so we need an explicit cast here otherwise we will run into problems when we enable `strict_types`.

Reply:
> From my understanding, the original problem was that `['default']` could be i.e. integer (when defined in ORM for example)

Yes.

> so we need an explicit cast here otherwise we will run into problems when we enable `strict_types`.

Does it apply to builtin functions? Anyways, my original implementation was done under the assumption that `strcmp(mixed $value1, mixed $value2)` is the same as `(string) $value1 === (string) $value2` which is not true. `strcmp()` expects strings.

Reply:
Yep it does: https://3v4l.org/BVrSK